corehq/apps/hqcase/case_helper.py

```diff
@@ -163,6 +167,110 @@ def close(
             is_creation=False,
         )
 
+    @classmethod
+    def copy_cases(
+        cls,
+        domain,
+        case_ids,
+        to_owner,
+        censor_data=None,
+        count_only=False
+    ):
+        """
+        Copies the cases governed by 'case_ids' to the respective 'to_owner'
+        and replacing the relevant case properties with the specified 'replace_props'
+        on the copied cases (useful for hiding sensitive data).
+
+        :param domain: the domain string
+        :param case_ids: case ids of the cases to copy
+        :param to_owner: the owner to copy the cases to
+        :param censor_data: the attributes/properties to be censored, specified as a dict.
+                The keys corresponding to the property/attribute and the value specifies
+                the type of censored data, i.e. number or date
+        :param count_only: specifies whether to return only the number of copied cases.
+
+        :return: The copied cases. If `count_only` is True only the count will be returned.
+        """
+        if not to_owner:
+            raise Exception('Must copy cases to valid new owner')
+
+        original_cases = CommCareCase.objects.get_cases(case_ids=case_ids, domain=domain)
+        if not any(original_cases):
+            return []
+
+        processed_cases = {}
+        copied_cases_case_blocks = []
+
+        def _get_new_identifier_index(case_identifier_index):
+            _identifier_index = {**case_identifier_index}
+            original_referenced_id = _identifier_index['case_id']
+
+            # We need the copied case's case_id for the new index
+            if original_referenced_id in processed_cases:
+                _identifier_index['case_id'] = processed_cases[original_referenced_id].case_id
+            else:
+                # Need to process the referenced case first to get the case_id of the copied case
+                try:
+                    original_parent_case = next((
+                        orig_case for orig_case in original_cases if orig_case.case_id == original_referenced_id
+                    ))
+                except StopIteration:
+                    return {}
+                referenced_case_block = _process_case(original_parent_case)
+                _identifier_index['case_id'] = referenced_case_block.case_id
+            return (
+                _identifier_index['case_type'],
+                _identifier_index['case_id'],
+                _identifier_index['relationship'],
+            )
+
+        def _get_new_index_map(_case):
+            new_case_index = {}
+            for identifier in _case.get_index_map():
+                identifier_index = _get_new_identifier_index(
+                    _case.get_index_map()[identifier]
+                )
+                if identifier_index:
+                    new_case_index[identifier] = identifier_index
+            return new_case_index
+
+        def _process_case(_case):
```

Comment:
This feature can be modelled as a Case Import?

Have you considered using the case-importer related code here directly instead of writing this afresh? This would help keep the case-modification logic in a single place making this, case-import and case-reassignment consistent

Reply:
Update: I will start looking into modelling this as a special purpose case import.

Reply:
@sravfeyn 
> Have you considered using the case-importer related code here directly instead of writing this afresh? This would help keep the case-modification

It looks like there's a lot that will need changing to make this compatible with the case importer code and/or vice versa. 
That said, I do think it might be helpful to have an abstract class somewhere that handles this type of operation which both this code and the importer code can simply use or inherit from, so I'll create a tech debt ticket to investigate that.